### ISSUE TITLE
Not clear behavior of Worker limiter & group.limit settings

### ISSUE THREAD

**Issue by dginzbourg** [taskforcesh/bullmq-pro-support#31]
Hi
I define Worker options as 
``` 
{
 concurrency: 100,
 limiter: {
    max: 2,
    duration: 1,
  },
  group: {
    limit: {
      max: 1,
      duration: 1,
      concurrency: 1,
    },
  } as WorkerProOptions,
}
```
then add 1 job with some group id. On the first job run and all the subsequent job processing, I throttle it with 
```
    await worker.rateLimitGroup(job, 1);
    throw WorkerPro.RateLimitError();
```
The job doesn't continue its run, I waited a couple of minutes.
If I set limiter.max = 1 or group.limit.max = 2 then the job processing is triggered within ~30ms. 
If I remove `limiter` completely then the job is triggered every 5s after been throttled. 
I don't understand this behavior. I thought the `limiter` is global for the worker regardless group settings, i.e. rate-limits across all groups and be expected significantly greater than group.limit. When job is throttled, then BullMQ applies delay for specific group. Is my assumption correct? Could you elaborate on the behaviors I experience for better understanding?
Thanks

**Comment by manast** [comment]
We are looking into this...

**Comment by manast** [comment]
We found an issue when using very low durations and high concurrency, a fix is on its way.
When you use "rateLimitGroup" you are overriding the current rate limit for that group, so the group rate limiter will not be used until the rate limiter expires, then it will use the one in the worker settings, you can think about it as a default rate limiter.
The global rate limiter will limit on top of the group rate limiter. So for example, you could have a group rate limiter that limits to 5 jobs per second per group, but a global rate limiter that limits to 10 jobs per second. That means that max 5 jobs from a given group will be processed in one second, but max 10 per second in total independent of which group they belong to.

**Comment by manast** [comment]
Please upgrade to version 5.1.4 and let me know if the issue persists.

**Comment by dginzbourg** [comment]
After the upgrade the current code stopped working. When job is rate limited, it isn't rescheduled. 5.1.1 works most of the time.

**Comment by manast** [comment]
Can you please paste a test code that is not working so that we can use it to reproduce the issue?

**Comment by dginzbourg** [comment]
We're using Nestjs + Nestjs BullMQ module, it'll take time to extract to test code, I'll do it on Sunday. I'll add  pseudo code just to check if your tests pass it. 

Queue/worker options:
```
{  
concurrency: 1000,
  limiter: {
    max: 1,
    duration: 1,
  },
  group: {
    limit: {
      max: 1,
      duration: 1,
      concurrency: 1,
    },
  } as WorkerProOptions
``` 
each job trigger the code looks like: 
```
   submitBatchOfItems(items); 
   if(moreItemsRemain()){
       await worker.rateLimitGroup(job, 1);
       throw WorkerPro.RateLimitError();
   }
```
What I'm doing is implementing BatchJob, each job trigger I Iterate over list of item, perform some batch operation on it, update job progress with the current list position and by delaying the job again achieve kind of repeatable job with high resolution of repeating. 
The worker options supposed to be 
```
{  
concurrency: 1000,
  limiter: {
    max: 1000,
    duration: 1000,
  },
  group: {
    limit: {
      max: 1,
      duration: 1,
      concurrency: 1,
    },
  } as WorkerProOptions
```   
but it didn't work. I could potentially increase delay on rateLimitGroup to ~10-50ms as well as group duration, but I'd like to achieve that each logical repetition will happen every 1s. 
On 5.1.4 the job wasn't scheduled after 1st delay.

**Comment by manast** [comment]
This is the test code I am using trying to reproduce your issue. This did not work before but works now on the latest version. Maybe you can tweak it so it is similar to your code so that it reproduces the issue:
```ts

import { QueuePro, WorkerPro } from "@taskforcesh/bullmq-pro";

const queueName = "pro-test";

const connection = { host: "localhost" };

const opts = {
  concurrency: 100,
  limiter: {
    max: 1,
    duration: 100,
  },
  group: {
    limit: {
      max: 1,
      duration: 100,
    },
  },
};

const queue = new QueuePro(queueName, { connection });

async function enqueueJob(groupId, payload) {
  await queue.add(queueName, payload, { group: { id: groupId } });
}

async function enqueueJobs() {
  const jobs = [];
  for (let g = 0; g < 5; g++) {
    for (let i = 0; i < 10; i++) {
      jobs.push(enqueueJob(g, { group: g, id: `g${g}-#${i}` }));
    }
  }
  return Promise.all(jobs);
}

export async function runQueueTest() {
  console.log(new Date().toISOString() + ": Adding jobs");
  await enqueueJobs();
  console.log(new Date().toISOString() + ": Jobs added");

  const worker = new WorkerPro(
    queueName,
    async (job) => {
      const id = job.data.id;
      const group = job.data.group;

      console.log(
        `${new Date().toISOString()}: enter ${job.data.id}`,
        job.attemptsMade
      );

      if (job.attemptsMade == 1) {
        await worker.rateLimitGroup(job, 1);
        throw WorkerPro.RateLimitError();
      }
    },
    { connection, ...opts }
  );
}

runQueueTest();
```

**Comment by dginzbourg** [comment]
I tried your code and it works for my settings. I'll debug my program on Sunday to analyze the behavior. Are you familiar with bull-repl that supports Pro groups stats? I'd like to get the visibility into the current queues/groups status without writing new code.

**Comment by manast** [comment]
Taskforce.sh supports groups for direct connections or use ngrok if you cannot use a direct connection. There is also a terminal based UI but it is not so nice.

**Comment by dginzbourg** [comment]
Hi. It seems that I'm able to reproduce the problem. I just added url call. In this settings, I'd expect 100 jobs runs in parallel and finish everything within ~1s, but it runs with ~8-10 jobs/sec. 

```ts
import { QueuePro, WorkerPro } from '@taskforcesh/bullmq-pro';
import got from 'got';

const queueName = 'pro-test';

const connection = { host: 'localhost', port: 6380, enableOfflineQueue: false };

const opts = {
  concurrency: 1000,
  limiter: {
    max: 100 * 1000,
    duration: 1000,
  },
  group: {
    limit: {
      max: 100,
      duration: 1000,
      concurrency: 100,
    },
  },
};

const queue = new QueuePro(queueName, { connection });

async function enqueueJob(groupId, payload) {
  await queue.add(queueName, payload, { group: { id: groupId } });
}

const groups = 1;
const members = 100;

async function enqueueJobs() {
  const jobs = [];
  for (let g = 0; g < groups; g++) {
    for (let i = 0; i < members; i++) {
      jobs.push(enqueueJob(g, { group: g, id: `g${g}-#${i}` }));
    }
  }
  return Promise.all(jobs);
}

export async function runQueueTest() {
  console.log(new Date().toISOString() + ': Adding jobs');
  await enqueueJobs();
  console.log(new Date().toISOString() + ': Jobs added');

  const worker = new WorkerPro(
    queueName,
    async (job) => {
      const id = job.data.id;
      const group = job.data.group;

      console.log(`${new Date().toISOString()}: En ${job.data.id}`, job.attemptsMade);
      const s = Date.now();
      const resp = await got('https://postman-echo.com/get');
      console.log(`${new Date().toISOString()}:   req time ${Date.now() - s} ${job.data.id}`, job.attemptsMade);

      if (job.attemptsMade < 1) {
        console.log(`${new Date().toISOString()}:  Th ${job.data.id}`, job.attemptsMade);
        await worker.rateLimitGroup(job, 10);
        throw WorkerPro.RateLimitError();
      }
      console.log(`${new Date().toISOString()}: Co ${job.data.id}`, job.attemptsMade);
    },
    { connection, ...opts }
  );
}

runQueueTest();
```

**Comment by manast** [comment]
I am not sure that what you get is incorrect. Keep in mind that every time you call ```worker.rateLimitGroup(job, 10);``` you are rate limiting the queue group 10ms. It would be better if you define what is it that you want to achieve and then we can figure out the best way to write the code for that.

**Comment by dginzbourg** [comment]
The original issue isn't reproducing with your code, it works with nestjs when NODE_ENV=test, but not 'production'. It works fine if I remove global concurrency. It's less a priority for now, I'll try to create nestjs based app to reproduce it.
Regarding my last comment, it's not related to delay, but the amount of scheduled jobs. I'll continue here, but if you'd like me to open separate issue, just tell me. 
For the settings:
```
  group: {
    limit: {
      max: 100,
      duration: 1000,
      concurrency: 100,
    }
```
I'd expect to have max 100 job scheduled within 1000ms regardless job time execution. If job exec time < 1s then I'd expect to see ~100 jobs every sec. In case some jobs are slow, then 'concurrency' parameter will limit next scheduling period.

**Comment by manast** [comment]
Please be aware that concurrency is not a property of limit but of group:
```ts
  group: {
    concurrency: 100,
    limit: {
      max: 100,
      duration: 1000,
    }
```

**Comment by dginzbourg** [comment]
Thanks for pointing out :). I wondered how it relates, but the exception defines that it's mutual exclusive with 'limit'.  The full settings for my last comment is 
```
const opts = {
  concurrency: 1000,
  limiter: {
    max: 100 * 1000,
    duration: 1000,
  },
  group: {
    limit: {
      max: 100,
      duration: 1000,
    },
  },
} as WorkerProOptions;
```

**Comment by manast** [comment]
I am not able to follow this issue anymore, it got too confusing...

**Comment by dginzbourg** [comment]
I'll open another issue to keep it around only the understanding of the  jobs execution speed

**Comment by dginzbourg** [comment]
I removed opt.limiter and kept only opt.concurrency and that solved the slowness I saw. I don't know why opt.limiter affects the run, but it works fine for me. 
I'm closing this issue, when and if I have small code example that reproduces the problem with delayed task not rescheduled I'll open another one.
Thank you for promptly responses, fixes & support.
